packages/dds/tree/src/test/shared-tree/sharedTree.bench.ts

```diff
@@ -0,0 +1,815 @@
+	rootFieldKeySymbol,
+	SchemaData,
+	TreeSchemaIdentifier,
+	UpPath,
+	ValueSchema,
+} from "../../core";
+
+/**
+ * Shapes of trees to test with.
+ */
+enum TreeShape {
+	/**
+	 * Each node is inserted directly 1 layer below the root node, making a wide tree.
+	 * Used to measure the performance of editing long sequences.
+	 */
+	Wide = 0,
+	/**
+	 * Each node is inserted one layer below the previous node, making a stick-like tree.
+	 * Used to measure the performance costs related to the length of paths.
+	 */
+	Deep = 1,
+}
+
+async function createSharedTrees(
+	schemaData: SchemaData,
+	data: JsonableTree[],
+	numberOfTrees = 1,
+): Promise<readonly [ITestTreeProvider, readonly ISharedTree[]]> {
+	const provider = await TestTreeProvider.create(numberOfTrees);
+	for (const tree of provider.trees) {
+		assert(tree.isAttached());
+	}
+	provider.trees[0].storedSchema.update(schemaData);
+	provider.trees[0].context.root.insertNodes(0, data.map(singleTextCursor));
+	await provider.ensureSynchronized();
+	return [provider, provider.trees];
+}
+
+const globalFieldKey: GlobalFieldKey = brand("foo");
+const localFieldKey: LocalFieldKey = brand("foo");
+const rootSchemaName: TreeSchemaIdentifier = brand("Test");
+
+function getTestSchema(fieldKind: { identifier: FieldKindIdentifier }): SchemaData {
+	const testRootNodeSchema = namedTreeSchema({
+		name: rootSchemaName,
+		localFields: {
+			[localFieldKey]: fieldSchema(fieldKind),
+		},
+		globalFields: [globalFieldKey],
+		value: ValueSchema.Serializable,
+	});
+	const testSchemaMap: Map<TreeSchemaIdentifier, NamedTreeSchema> = new Map();
+	testSchemaMap.set(rootSchemaName, testRootNodeSchema);
+	testSchemaMap.set(dataSchema.name, dataSchema);
+	return {
+		treeSchema: testSchemaMap,
+		globalFieldSchema: new Map([
+			[rootFieldKey, fieldSchema(FieldKinds.value, [rootSchemaName])],
+		]),
+	};
+}
+
+const dataSchema = namedTreeSchema({
+	name: brand("DataSchema"),
+	localFields: {
+		foo: fieldSchema(FieldKinds.optional),
+	},
+	value: ValueSchema.Number,
+});
+
+// number of nodes in test for wide trees
+const nodesCountWide = [
+	[1, BenchmarkType.Measurement],
+	[100, BenchmarkType.Perspective],
+	[500, BenchmarkType.Measurement],
+];
+// number of nodes in test for deep trees
+const nodesCountDeep = [
+	[1, BenchmarkType.Measurement],
+	[10, BenchmarkType.Perspective],
+	[100, BenchmarkType.Measurement],
+];
+
+const rootFieldSchema = fieldSchema(FieldKinds.value);
+const globalFieldSchema = fieldSchema(FieldKinds.value);
+const rootNodeSchema = namedTreeSchema({
+	name: brand("TestValue"),
+	localFields: {
+		optionalChild: fieldSchema(FieldKinds.optional, [brand("TestValue")]),
+	},
+	extraLocalFields: fieldSchema(FieldKinds.sequence),
+	globalFields: [globalFieldKey],
+});
+const testSchema: SchemaData = {
+	treeSchema: new Map([[rootNodeSchema.name, rootNodeSchema]]),
+	globalFieldSchema: new Map([
+		[rootFieldKey, rootFieldSchema],
+		[globalFieldKey, globalFieldSchema],
+	]),
+};
```

Comment:
I'm unclear why you have the schema setup this way, but some parts of this are not valid.

testSchema only contains one type (rootNodeSchema), so thats the only type allowed in a tree using that schema. However rootNodeSchema references a type named "TestValue" which you never define. It seems like maybe you meant "DataSchema"? Also "DataSchema" is not included.

Reply:
updated schema